kapitan/inputs/base.py

```diff
@@ -0,0 +1,147 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import errno
+import logging
+import os
+import yaml
+import ujson as json
+
+from kapitan.errors import CompileError, KapitanError
+from kapitan.refs.base import Revealer
+from kapitan.utils import PrettyDumper
+
+logger = logging.getLogger(__name__)
+
+
+class InputType(object):
+    def __init__(self, type_name, compile_path, search_paths, ref_controller):
+        self.type_name = type_name
+        self.compile_path = compile_path
+        self.search_paths = search_paths
+        self.ref_controller = ref_controller
+
+    def compile_obj(self, comp_obj, ext_vars, **kwargs):
+        """
+        run compile_input_path() for each input_path in comp_obj
+        kwargss are passed into compile_input_path()
+        """
+        input_type = comp_obj["input_type"]
+        assert input_type == self.type_name
+        input_paths = comp_obj["input_paths"]
+
+        for input_path in input_paths:
+            self.compile_input_path(input_path, comp_obj, ext_vars, **kwargs)
+
+    def compile_input_path(self, input_path, comp_obj, ext_vars, **kwargs):
+        """
+        compile and validate input_path in comp_obj
+        kwargs are passed into compile_file()
+        """
+        target_name = ext_vars["target"]
+        output_path = comp_obj["output_path"]
+        output_type = comp_obj.get("output_type", self.default_output_type())
+        file_found = False
+
+        for path in self.search_paths:
+            compile_file_sp = os.path.join(path, input_path)
+            if os.path.exists(compile_file_sp):
+                file_found = True
+                logger.debug("Compiling %s", compile_file_sp)
+                try:
+                    _compile_path = os.path.join(self.compile_path, target_name, output_path)
+                    self.compile_file(compile_file_sp, _compile_path, ext_vars, output=output_type,
+                                      target_name=target_name, **kwargs)
+                except KapitanError as e:
+                    raise CompileError("{}\nCompile error: failed to compile target: {}".format(e, target_name))
+
+        if not file_found:
+            raise CompileError("Compile error: {} for target: {} not found in "
+                               "search_paths: {}".format(input_path, target_name, self.search_paths))
+
+    def make_compile_dirs(self, target_name, output_path):
+        """make compile dirs, skips if dirs exist"""
+        _compile_path = os.path.join(self.compile_path, target_name, output_path)
+        # support writing to an already existent dir
+        try:
+            os.makedirs(_compile_path)
```

Comment:
Use `os.makedirs(_compile_path, exist_ok=True)` then you can remove the try-catch block you have in case it fails.

Reply:
Good point. Done